snippets/ObjectDestroy.hpp

```diff
@@ -1,64 +1,139 @@
+  }
+
+  Dispatch const & getDispatch() const VULKAN_HPP_NOEXCEPT
+  {
+    return *m_dispatch;
+  }
+
+protected:
+  template <typename T>
+  void destroy( T t ) VULKAN_HPP_NOEXCEPT
+  {
+    VULKAN_HPP_ASSERT( m_dispatch );
+    t.destroy( m_allocationCallbacks, *m_dispatch );
+  }
+
+private:
+  Optional<const AllocationCallbacks> m_allocationCallbacks = nullptr;
+  Dispatch const *                    m_dispatch            = nullptr;
+};
+
+//================================================================================================
+
+template <typename HandleType>
+class ObjectDestroyShared
+{
+public:
+  using handle_type = HandleType;
+  using parent      = parent_of_t<handle_type>;
+
+  template <class Dispatcher>
+  using destroy_pfn_t =
+    typename std::conditional<has_parent<handle_type>,
+
+                              void ( parent::* )( handle_type kty, const AllocationCallbacks * pAllocator, const Dispatcher & d ) const VULKAN_HPP_NOEXCEPT,
+
+                              void ( handle_type::* )( const AllocationCallbacks * pAllocator, const Dispatcher & d ) const VULKAN_HPP_NOEXCEPT>::type;
+
+  ObjectDestroyShared() = default;
+
+  template <typename Dispatcher, typename = typename std::enable_if<has_parent<handle_type>>::type>
+  ObjectDestroyShared( Optional<const AllocationCallbacks> allocationCallbacks VULKAN_HPP_DEFAULT_ARGUMENT_NULLPTR_ASSIGNMENT,
+                       const Dispatcher & disp                                 VULKAN_HPP_DEFAULT_DISPATCHER_ASSIGNMENT )
+    : m_destroy( reinterpret_cast<decltype( m_destroy )>( static_cast<destroy_pfn_t<Dispatcher>>( &parent::destroy ) ) )
+    , m_loader( &disp )
+    , m_allocationCallbacks( allocationCallbacks )
+  {
+  }
+
+  template <typename Dispatcher, typename = typename std::enable_if<!has_parent<handle_type>>::type>
+  ObjectDestroyShared( Optional<const AllocationCallbacks> allocationCallbacks VULKAN_HPP_DEFAULT_ARGUMENT_NULLPTR_ASSIGNMENT,
+                       const Dispatcher & disp                                 VULKAN_HPP_DEFAULT_DISPATCHER_ASSIGNMENT )
+    : m_destroy( reinterpret_cast<decltype( m_destroy )>( static_cast<destroy_pfn_t<Dispatcher>>( &handle_type::destroy ) ) )
+    , m_loader( &disp )
+    , m_allocationCallbacks( allocationCallbacks )
+  {
+  }
+
+protected:
+  std::enable_if<has_parent<handle_type>, void>::type Destroy( parent parent, handle_type handle ) const VULKAN_HPP_NOEXCEPT
+  {
+    VULKAN_HPP_ASSERT( m_destroy && m_loader );
+    ( parent.*m_destroy )( handle, m_allocationCallbacks, *m_loader );
+  }
+
+  std::enable_if<!has_parent<handle_type>, void>::type Destroy( handle_type handle ) const VULKAN_HPP_NOEXCEPT
+  {
+    VULKAN_HPP_ASSERT( m_destroy && m_loader );
+    ( handle.*m_destroy )( m_allocationCallbacks, *m_loader );
+  }
+
+private:
+  destroy_pfn_t<DispatchLoaderBase>   m_destroy             = nullptr;
+  const DispatchLoaderBase *          m_loader              = nullptr;
```

Comment:
I could use any refence. The problem lies deeper: shared handles need to track dispatchers of their parents, and if you have parent with different dispatcher, then standard one, you will have a problem removing, since you lose the guarantee of correct destruction by the dispatcher. 
This was not the issue with original Wisdom `shared_handle`, since I always use dynamic dispatcher and didn't need to bother. 
So I provide a templated constructor to capture `destroy_pfn_t<OriginalDisp>` and ref to original dispatcher, to later invoke it as a binary compatible (mostly forcefully). This is the way to avoid using `std::bind()` or `std::function`, since they are too much overhead.

Reply:
> if you have parent with different dispatcher

You mean, if you have parents with different dispatchers **in the same executable**? Like having parents using `vk::DispatchLoaderStatic`, and others using `vk::DispatchLoaderDynamic`?

Reply:
correct